### DISCUSSION TITLE
grpc - http/json to grpc transcoding

### DISCUSSION THREAD

**Discussion by zZHorizonZz** [quarkusio/quarkus#39991] [Q&A]
I'm currently working on a gRPC transcoding implementation for Quarkus. It's in the prototype stage, but I'd like to possibly contribute it as an official Quarkus feature. I'm currently looking for some feedback on the implementation and overall direction. One specific question I have is about endpoint registration. Currently, I'm using resteasy by adding @ApplicationScoped to the class, but a comment in this thread (https://github.com/quarkusio/quarkus/issues/25837#issuecomment-1144528115) suggests registering it via the router directly, which I can't wrap my head around on how should I implement it. The repo can be found [here](https://github.com/zZHorizonZz/quarkus-grpc-transcoding).

[a comment by an automated account is not shown]

**Reply by cescoffier** [reply]
That's cool!

I would register the routes on the router.

Note that @tsegismont has recently worked on support for gRPC Web.

**Reply by tsegismont** [reply]
Right, this new feature will be added to Vert.x 5, still not sure about backporting to Vert.x 4

**Reply by cescoffier** [reply]
It should be considered for Vert.x 4 - Vert.x 5 is quite far for Quarkus (like years (plural)  far)

**Reply by zZHorizonZz** [reply]
> Note that @tsegismont has recently worked on support for gRPC Web.

So, if I understand correctly, this would imply native support for the transcoding of http/json to gRPC based on protobuf specifications? If that's the case, that's very cool!

**Reply by tsegismont** [reply]
No, it would imply native support for the [gRPC-Web protocol](https://github.com/grpc/grpc-web).

**Reply by zZHorizonZz** [reply]
I see that's quite interesting. So, since it works on HTTP, is the structure of a request the same as if I send the request to a transcoded endpoint, like this JSON:

```json
{
  "name": "test"
}
```

Which would be based on this message:

```proto
message TestRequest {
 string name = 1;
}
```

Or, since it's called protocol, is the data sent in its own unique format?

**Reply by tsegismont** [reply]
gRPC-Web has its own [wire format](https://github.com/grpc/grpc-web?tab=readme-ov-file#wire-format-mode), which works on both HTTP/1.1 and HTTP/2.

**Reply by zZHorizonZz** [reply]
@cescoffier 

> I would register the routes on the router.

Yes, that approach sounds better. However, I'm not sure about how to best execute the method itself. Currently, I can generate a class using resteasy to create endpoints. My initial idea was to reuse this functionality and call the methods directly from the router. The downside is that this bypasses interceptors and other grpc features. While this is ok for my personal project, If I want to contribute this to  Quarkus I think I should try to find a better approach maybe something like mirroring the execution approach used by `GrpcServerImpl`. Unfortunately, I can't directly reuse `GrpcServerImpl` since it manipulates the HTTP response in a way I can't modify (Or I didn't find a way on how to modify it). I'm considering custom `ServerRequestImpl`, `ServerResponseImpl`, and `GrpcServer` implementations for transcoding, but I'm not sure if that's the best approach. I'd greatly appreciate any ideas or suggestions you have.

**Reply by cescoffier** [reply]
Yeah, that's why the Vert.x approach is "simpler" as they are at the server level. In your case, you will have to open a few things in the server impl to be able to invoke the services.

**Comment by alesj** [comment]
@zZHorizonZz I once wrote this proxy code, if it helps
* https://github.com/quarkusio/quarkus/pull/30777

You would just need to add `json <--> gRPC` decoder / encoder ...
